java/jucx-binding/src/main/java/de/hhu/bsinfo/observatory/jucx/JucxBenchmark.java

```diff
@@ -0,0 +1,237 @@
+                .setPeerErrorHadnlingMode();
+        clientToServer = worker.newEndpoint(epParams);
+        // Exchange small message to wire-up connection
+        ByteBuffer buffer = ByteBuffer.allocateDirect(8);
+        worker.progressRequest(clientToServer.sendTaggedNonBlocking(buffer, null));
+        worker.progressRequest(worker.recvTaggedNonBlocking(buffer, null));
+        return Status.OK;
+    }
+
+    private void exchangeMemoryInformation() {
+        UcpEndpoint endpoint = (clientToServer != null) ? clientToServer : serverToClient;
+        ByteBuffer recvMemoryRkey = recvMemory.getRemoteKeyBuffer();
+        ByteBuffer sendMessage = ByteBuffer.allocateDirect(8 + recvMemoryRkey.capacity());
+        ByteBuffer recvMessage = ByteBuffer.allocateDirect(4096);
+
+        sendMessage.putLong(recvMemory.getAddress());
+        sendMessage.put(recvMemoryRkey);
+        sendMessage.clear();
+
+        UcpRequest sendRequest = endpoint.sendStreamNonBlocking(sendMessage, null);
+        UcpRequest recvRequest = endpoint.recvStreamNonBlocking(recvMessage, 0L, null);
+
+        worker.progressRequest(sendRequest);
+        worker.progressRequest(recvRequest);
+
+        remoteAddress = recvMessage.getLong();
+        remoteKey = endpoint.unpackRemoteKey(recvMessage);
+
+        resources.add(endpoint);
+        resources.add(remoteKey);
+    }
+
+    @Override
+    protected Status prepare(int operationSize) {
+        LOGGER.info("Preparing memory regions and exchanging metadata");
+        sendMemory = context.registerMemory(ByteBuffer.allocateDirect(operationSize));
+        recvMemory = context.registerMemory(ByteBuffer.allocateDirect(operationSize));
+
+        resources.add(sendMemory);
+        resources.add(recvMemory);
+
+        exchangeMemoryInformation();
+        return Status.OK;
+    }
+
+    @Override
+    protected Status cleanup() {
+        LOGGER.info("Cleanup ...");
+        while (!resources.isEmpty()) {
+            try {
+                resources.pop().close();
+            } catch (IOException e) {
+                return Status.FILE_ERROR;
+            }
+        }
+        return Status.OK;
+    }
+
+    @Override
+    protected Status fillReceiveQueue() {
+        LOGGER.info("Posting recv request");
+        worker.recvTaggedNonBlocking(recvMemory.getAddress(), recvMemory.getLength(), 0, 0, null);
+        return Status.OK;
+    }
+
+    @Override
+    protected Status sendMultipleMessages(int messageCount) {
+        LOGGER.info("Sending {} messages", messageCount);
+        UcpRequest[] requests = new UcpRequest[messageCount];
+        UcpEndpoint endpoint = (clientToServer != null) ? clientToServer : serverToClient;
+
+        for (int i = 0; i < messageCount; i++) {
+            requests[i] = endpoint.sendTaggedNonBlocking(sendMemory.getAddress(),
+                    sendMemory.getLength(), 0, null);
+        }
+
+        while (Arrays.stream(requests).allMatch(UcpRequest::isCompleted)) {
```

Comment:
Maybe consider using a loop instead of the Streaming API, as a simple loop would likely be faster and create no garbage. However, this is just a suggestion and you probably know best how to achieve the best performance with UCX.